src/compiler/checker.ts

```diff
@@ -36286,6 +36286,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const nextTypes: Type[] = [];
         const isAsync = (getFunctionFlags(func) & FunctionFlags.Async) !== 0;
         forEachYieldExpression(func.body as Block, yieldExpression => {
+            const statement = findAncestor(yieldExpression, isStatement)!;
+            if (canHaveFlowNode(statement) && !statement.flowNode && !compilerOptions.allowUnreachableCode) {

```

Comment:
Out of curiousity, why flag this on `allowUnreachableCode`? My impression is that `allowUnreachableCode` is more like a lint rule that errors on the unreachable code, and not something that should impact type analysis.

Reply:
I honestly have no idea cause I've written it half a year ago 😅 

If the tests would come roughly unchanged without this condition (which is likely) then maybe I just have tried to be conservative when it comes to the impact of this change. Taking a look at how `allowUnreachableCode` is defined in the docs now - you are right that this flag should mostly just affect error reporting.